lib/model-builder.js

```diff
@@ -596,14 +597,17 @@ ModelBuilder.prototype.define = function defineClass(className, properties, sett
           } else {
             if (DataType === List) {
               this.__data[propertyName] = isClass(DataType) ?
-                new DataType(value, properties[propertyName].type, this.__data) :
-                DataType(value, properties[propertyName].type, this.__data);
+                new DataType(value, properties[propertyName].type, this) :
+                DataType(value, properties[propertyName].type, this);
```

Comment:
I am concerned about this change. A model instance contains additional properties on top of what's stored in `__data`, for example it can contain cached relations. I am reluctant to change the contract of `DataType` constructor in this way.

How about adding a 4th argument? That way we preserve backwards compatibility for any existing data-types that may relay on `this.__data` behavior.

```suggestion
                DataType(value, properties[propertyName].type, this.__data, this);
```

(This comment applies to both changed lines.)

Reply:
I could also add a `__parent` property in the `List` instance, as it is with the rest instead of tampering with the default `parent` if that suits you

Reply:
Will implement as discussed

Reply:
I am afraid my comment haven't been addressed yet.

I think it's important to keep sending `this.__data` to the second argument of `DataType` constructor. As I understand it, `DataType` can be possibly a custom, user-provided function/constructor. Such functions may not be prepared to receive full model instance instead of plain model data.

Reply:
I accidentally forgot to reply on this conversation. From what I can tell the `List` is currently used except for this commented line in other two places with the parent feature usage:

1. https://github.com/strongloop/loopback-datasource-juggler/blob/ce788c54d20c0e4bf16f36eec627fc60b2445a4e/lib/dao.js#L1670
2. https://github.com/strongloop/loopback-datasource-juggler/blob/ce788c54d20c0e4bf16f36eec627fc60b2445a4e/lib/model.js#L362

And also without parent usage in:
https://github.com/strongloop/loopback-datasource-juggler/blob/ce788c54d20c0e4bf16f36eec627fc60b2445a4e/lib/include.js#L1009

The current usages *(possibly by mistake)*, do pass on to the `List` constructor the actual model instance and not the `__data` property as the discussed property setter does. So this is how it has been used all this time since at least the code in `Model._initProperties` is what is run on every db query etc.
That is why I initially changed the setter property behavior to comply with existing usages since it was the only different.

So we either leave this as it is, or change the other usages as well to the new behavior *(first accepting a plain object instead of model) in order to achieve final consistency.

Please advise on how you want me to proceed

Reply:
Thank you for the detailed explanation! My main concern was about breaking non-List use cases, because I did not realize this code is already inside `if (DataType === List) {` block.

Based on your description of the current state, and the fact that this change affects Lists only, I am fine to use your current proposal :+1: